torch/_inductor/autotune_process.py

```diff
@@ -131,15 +127,121 @@ def close(self) -> None:
+        if self.process is not None:
             self.process.wait()
             self.close()
 
 
-tuning_process = TuningProcess()
+@dataclasses.dataclass
+class TuningProcessPool:
+    processes: Optional[Queue[TuningProcess]] = None
+    executor: Optional[ThreadPoolExecutor] = None
+
+    def initialize(self, count: Optional[int] = None) -> None:
+        """
+        Start the child processes.
+        """
+        assert (self.processes is None) == (self.executor is None)
+        if self.processes is not None:
+            return
+
+        count = count or torch.cuda.device_count()
+        assert count > 0
+
+        # Launch the child processes and push a msg to "warm up"
+        self.processes = Queue()
+        for device in range(count):
+            p = TuningProcess(device=device)
+            p.initialize()
+            p.put(Ping())
+            self.processes.put(p)
+
+        # Wait for the initialization to finish
+        for p in self.processes.queue:
+            assert isinstance(p.get(), Pong)
+
+        # Use a thread pool to manage distributing work to the subprocesses.
+        # Threads block on an available process, so it makes sense to match
+        # the number of threads with the number of devices.
+        self.executor = ThreadPoolExecutor(max_workers=count)
+
+        # Register the exit handler for the parent process so it will terminate
+        # the child processes.
+        global EXIT_HANDLER_REGISTERED
+        if not EXIT_HANDLER_REGISTERED:
+            EXIT_HANDLER_REGISTERED = True
+            import atexit
+
+            atexit.register(lambda: self.terminate())
+
+    def terminate(self) -> None:
+        """
+        Signal all child processes to terminate.
+        """
+        if self.executor is not None:
+            self.executor.shutdown()
+            self.executor = None
+
+        if self.processes is not None:
+            for p in self.processes.queue:
+                p.terminate()
+            for p in self.processes.queue:
+                p.wait()
+            self.processes = None
+
+    def target(self, choice: TritonTemplateCaller) -> float:
+        """
+        Entry point for the thread-pool helper threads: Wait for an open TuningProcess,
+        remove it from the queue, execute the benchmark in that subprocess, and return
+        the TuningProcess to the queue.
+        """
+        assert choice.bmreq is not None
+        assert self.processes is not None
+
+        process = self.processes.get()
+        process.put(choice.bmreq)
+        try:
+            return process.get()
+        except EOFError:
+            warnings.warn(
+                f"Failed to benchmark choice '{choice}'. It will be ignored. "
+                "Please debug the root cause in case the choice can bring perf gains."
+            )
+            # set to INF so this choice will be ignored
+            return float("inf")
+        finally:
+            self.processes.put(process)
```

Comment:
is there any worry here about putting back in the same process after it had failed ? say, if the process has been corrupted ? might be safer to just exit and start new process. what does old code do ?

Reply:
@eellison this is related to a comment you made on the parent diff. The way I tried to organize things, it's the TuningProcess class that manages its subprocess and always ensures that it's running. So it should be safe as long as the TuningProcess is written correctly to do that.

The old code always calls "initialize" from the outside the TuningProcess before using it, and it's that initialize method that checks if the process previously crashed and restarts it if so. I think the abstraction is cleaner if we just contain that logic inside the TuningProcess. One suggestion I made on the other diff was that it might be clearer to the reader if we optionally restart sub-processes before sending a message (instead, I did it immediately after crash). Either approach is fine, I think.

Reply:
Went with lazy per conversation on that line of code